foundations/src/telemetry/tokio/mod.rs

```diff
@@ -0,0 +1,364 @@
+//! | tokio_runtime_worker_parks_total                 | [`tokio::runtime::RuntimeMetrics::worker_park_count`]               | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_noops_total                 | [`tokio::runtime::RuntimeMetrics::worker_noop_count`]               | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_task_steals_total           | [`tokio::runtime::RuntimeMetrics::worker_steal_count`]              | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_steal_operations_total      | [`tokio::runtime::RuntimeMetrics::worker_steal_operations`]         | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_polls_total                 | [`tokio::runtime::RuntimeMetrics::worker_poll_count`]               | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_busy_duration_micros_total  | [`tokio::runtime::RuntimeMetrics::worker_total_busy_duration`]      | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_local_schedules_total       | [`tokio::runtime::RuntimeMetrics::worker_local_schedule_count`]     | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_overflows_total             | [`tokio::runtime::RuntimeMetrics::worker_overflow_count`]           | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_local_queue_depth           | [`tokio::runtime::RuntimeMetrics::worker_local_queue_depth`]        | runtime_name?, runtime_id?, worker_idx |
+//! | tokio_runtime_worker_mean_poll_time_micros       | [`tokio::runtime::RuntimeMetrics::worker_mean_poll_time`]           | runtime_name?, runtime_id?, worker_idx |
+//!
+//! ## Example
+//! ```no_run
+//! # use std::thread;
+//! # use std::time::Duration;
+//! # use foundations::telemetry::tokio::RuntimeMonitor;
+//! // create and monitor 8 runtimes spawned on other threads, then poll in the background from this one
+//! let mut monitor = RuntimeMonitor::new();
+//!
+//! for i in 0..8 {
+//!     let r = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();
+//!     
+//!     monitor.add_runtime(None, Some(i), r.handle());
+//!
+//!     thread::spawn(move || r.block_on(async {
+//!         loop {
+//!             println!("Poll");
+//!             tokio::time::sleep(Duration::from_secs(1)).await;
+//!         }
+//!     }));
+//! }
+//!
+//! loop {
+//!     monitor.record_sample();
+//!
+//!     // record metrics roughly twice a second
+//!     std::thread::sleep(Duration::from_secs_f32(0.5));
+//! }
+//! ```
+
+use crate::telemetry::metrics::{metrics, Counter, Gauge};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use tokio::runtime::{Handle, RuntimeMetrics};
+
+/// Monitors a set of runtimes, allowing metrics to be periodically collected for each tracked runtime.
+#[derive(Default)]
+pub struct RuntimeMonitor {
+    runtimes: Vec<RuntimeHandle>,
```

Comment:
It doesn't feel like we need to have multiple runtimes in a single monitor. It would have be useful if monitor was a global singleton, but it's not, so users can just create a monitor per runtime if they need to.

However, I personally would prefer a global singleton (metrics are global anyway) approach (MemoryProfiler could be an inspiration). Also, it probably would be nice to auto sampling built-in with given frequency and providing a function to enforce sampling manually

Reply:
I think a global singleton monitor approach is a good idea here.

We definitely _do_ need to support multiple runtimes for this API, especially if it is singleton.

Reply:
I've pushed up a static version.

Now, let's bikeshed the new names!